requests/models.py

```diff
@@ -455,28 +443,41 @@ def prepare_body(self, data, files, json=None):
                     else:
                         content_type = 'application/x-www-form-urlencoded'
 
-            self.prepare_content_length(body)
-
             # Add content-type if it wasn't explicitly provided.
             if content_type and ('content-type' not in self.headers):
                 self.headers['Content-Type'] = content_type
 
+        self.prepare_content_length(body)
         self.body = body
 
     def prepare_content_length(self, body):
-        if hasattr(body, 'seek') and hasattr(body, 'tell'):
-            curr_pos = body.tell()
-            body.seek(0, 2)
```

Comment:
No, it's handled differently but better. Regardless, it needs to be reconciled: we were inconsistently using either this logic or `super_len` prior to this patch, which means that in unlikely scenarios we could get different answers to the content-length of the same thing.